src/libasr/pass/intrinsic_function_registry.h

```diff
@@ -453,6 +455,54 @@ namespace Abs {
 
 } // namespace Abs
 
+namespace ListIndex {
+
+static inline ASR::expr_t *eval_list_index(Allocator &/*al*/,
+    const Location &/*loc*/, Vec<ASR::expr_t*>& /*args*/) {
+    // TODO: To be implemented for ListConstant expression
+    return nullptr;
+}
+
+static inline ASR::asr_t* create_ListIndex(Allocator& al, const Location& loc,
+    Vec<ASR::expr_t*>& args,
+    const std::function<void (const std::string &, const Location &)> err) {
+    if (args.size() != 2) {
+        // Support start and end arguments by overloading ListIndex
+        // intrinsic. We need 3 overload IDs,
+        // 0 - only list and element
+        // 1 - list, element and start
+        // 2 - list, element, start and end
+        // list, element and end case is not possible as list.index
+        // doesn't accept keyword arguments
+        err("For now index() takes exactly one argument", loc);
+    }
+
+    ASR::expr_t* list_expr = args[0];
+    ASR::ttype_t *type = ASRUtils::expr_type(list_expr);
+    ASR::ttype_t *list_type = ASR::down_cast<ASR::List_t>(type)->m_type;
+    ASR::ttype_t *ele_type = ASRUtils::expr_type(args[1]);
+    if (!ASRUtils::check_equal_type(ele_type, list_type)) {
+        std::string fnd = ASRUtils::type_to_str_python(ele_type);
+        std::string org = ASRUtils::type_to_str_python(list_type);
+        err(
+            "Type mismatch in 'index', the types must be compatible "
+            "(found: '" + fnd + "', expected: '" + org + "')", loc);
+    }
+    Vec<ASR::expr_t*> arg_values;
+    arg_values.reserve(al, args.size());
+    for( size_t i = 0; i < args.size(); i++ ) {
+        arg_values.push_back(al, ASRUtils::expr_value(args[i]));
+    }
+    ASR::expr_t* compile_time_value = eval_list_index(al, loc, arg_values);
+    ASR::ttype_t *to_type = ASRUtils::TYPE(ASR::make_Integer_t(al, loc,
+                            4, nullptr, 0));
+    return ASR::make_IntrinsicFunction_t(al, loc,
+            static_cast<int64_t>(ASRUtils::IntrinsicFunctions::ListIndex),
+            args.p, args.size(), 0, to_type, compile_time_value);
```

Comment:
Is the idea that `create_ListIndex` is to be used when we need to check the argument types and compute compile time value (if available), but if we already have everything ready (such as in some transformation passes), we just use make_IntrinsicFunction_t directly?

Reply:
Yes. All the error checking related to an intrinsic function resides in one place. Also there is no restriction on using `make_IntrinsicFunction_t` directly (its a public API and if you few feel you can avoid calling, `create_ListIndex` or any other create function then feel free to do so).

Reply:
Later we might need to add some checks to verify() to ensure all arguments to IntrinsicFunctions are correct. For now this is good.

Reply:
We can easily do this by registering `verify_args` function in `IntrinsicFunctionRegistry`. We can define, `LogGamma::verify_args`, `Abs::verify_args`, etc. Then, in asr_verify.cpp we can call these methods in the same way as we call `instantiate_function` in `intrinsic_function.cpp`.